### DISCUSSION TITLE
Can Saleor use SQS as a Celery Broker?

### DISCUSSION THREAD

**Discussion by dgurram** [saleor/saleor#8707] [Deployment]
Hi Saleor team,

We're trying to deploy Saleor 3.0 onto ECS Fargate on AWS with the DB being PostgreSQL on Aurora RDS. One of the configs required is the CELERY_BROKER_URL which I presume is used by the Worker. I'm wondering what I should set this to for the AWS environment. Do I need to set up an Elasticache Redis cluster? The Celery documentation does warn about potential data loss if using Redis however.

Celery also supports SQS - https://docs.celeryproject.org/en/stable/getting-started/backends-and-brokers/sqs.html#broker-sqs - but is the `celery[sqs]` installed by Saleor? Would it actually work?

Has anyone successfully got Saleor's worker working with SQS?

At a more general level, is the Worker an optional component or is it considered a part of Saleor Core? I don't see it mentioned on the [Architecture page](https://docs.saleor.io/docs/3.0/developer/getting-started/architecture). Is it only required for the cases mentioned [here](https://docs.saleor.io/docs/3.0/developer/running-saleor/background-tasks)? These wouldn't apply to us so perhaps I don't need to worry about deploying it?

Thanks,

Dinesh

**Comment by patrys** [comment]
You should be able to use any broker supported by Celery, we don't make any special assumptions about the broker or the transport. You may need to install additional dependencies depending on the broker you want to use.

The Celery worker is required for web hooks to work. Here's a Docker Compose example that overrides the command to start the Celery worker: https://github.com/saleor/saleor-platform/blob/c3e993878485cca7c0c55ea642ac374dc1ad890f/docker-compose.yml#L78-L96

**Reply by daltonsena** [reply]
@dgurram To use SQS as a broker in Celery, you need to install **pycurl**.

You could simple add SQS as an extra for celery dependency at the **pyproject.toml** file, something like:
```
    [tool.poetry.dependencies.celery]
    version = ">=4.4.5,<6.0.0"
    extras = [ "redis", "sqs" ]
```

If fits to you, you could remove the Redis reference, something like:
```
    [tool.poetry.dependencies.celery]
    version = ">=4.4.5,<6.0.0"
    extras = [ "sqs" ]
```

Also, taking this path, you will probably need to make some changes at the Dockerfile to not get errors during the image build (regarding the pycurl dependency), something like:
```
RUN apt-get update \
  && apt-get install -y \
  libcairo2 \
  libgdk-pixbuf2.0-0 \
  liblcms2-2 \
  libopenjp2-7 \
  libpango-1.0-0 \
  libpangocairo-1.0-0 \
  libssl1.1 \
  libtiff5 \
  libwebp6 \
  libxml2 \
  libpq5 \
  shared-mime-info \
  mime-support \
  libssl1.1 \
  curl \
  openssl \
  libcurl4-openssl-dev \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*
```

**libssl1.1**, **curl**, **openssl** and **libcurl4-openssl-dev** are new entries there.

Finally, something like this will put you in the game:

```
CELERY_BROKER_URL = (
    "sqs://{access_key}:{secret_key}@".format(
        access_key=quote(SQS_ACCESS_KEY_ID, safe=""),
        secret_key=quote(SQS_SECRET_ACCESS_KEY, safe=""),
    )
    or ""
)
CELERY_ACCEPT_CONTENT = ["application/json"]
CELERY_TASK_SERIALIZER = "json"
CELERY_RESULT_SERIALIZER = "json"
```

I added the serializers in this example because, if I remember right, SQS doesn't support another type of serialization, so If you have an external agent publishing messages to the queue (like a lambda, etc...), you need to pay attention to this.

[Look the CAVEATS section on the docs is highly recommended](https://docs.celeryproject.org/en/stable/getting-started/backends-and-brokers/sqs.html#caveats)

**Reply by dgurram** [reply]
Thanks @patrys  and @daltonsena for the details. Very helpful.

I was hoping to not change Saleor source but looks like at least the `pyproject.toml` and `Dockerfile` will need updates. Certainly manageable, but taking a step back - @patrys - is the worker only needed for webhooks? Does it have any other purpose? To begin with, we're not going to need web hooks so I'm thinking we might be able to make use of Saleor without the worker.

**Reply by korycins** [reply]
@dgurram I don't think that you can skip a worker. We are using it for more things than only the webhooks: emails, cleaning, async update after some action, etc. (You can just search ".delay" in saleor code to see all examples of celery usage).

**Reply by dgurram** [reply]
Thanks all for the help so far. Understood that the worker is important. We'll try out the suggestions above on configuring for SQS and come back with more info on how we fared.
